feat: 검색기능 구현(#27)

SowonHub opened the pull request:
## 📝 작업 개요

- 커뮤니티 목록페이지 검색기능을 검색 필터와 정렬을 같이 적용

## 📌 작업 상세 내용

- [x] useCommunityQuery, useCommunityPosts
- [x] filterPostsClient
- [x] 관련 이슈: #27 

## 🔍 체크 리스트

- [x] 디자인/플로우 문서와 구현 내용이 일치합니다.
- [ ] `npm run lint` 실행 시 에러/경고가 없습니다.
- [ ] `npm test` 실행 시 실패하는 테스트가 없습니다.
- [ ] 비회원 / 회원 / 작성자 플로우를 브라우저에서 직접 확인했습니다.

SowonHub committed:
remove -dom

SowonHub committed:
add routes

SowonHub committed:
edit simple

SowonHub committed:
add api community

SowonHub committed:
add hooks communityposts

SowonHub committed:
add hooks communityquery

SowonHub committed:
add utils community : category, search, search filter, sort

SowonHub committed:
add type category:string

SowonHub committed:
editing function search ( + filter, array )

redfoot135 reviewed (commented):
**시도·고민한 점 (Optional)**

- 한번에 다양한 기능이 가능하도록 상태관리를 어떻게 해야할지 고민중에 있습니다.

**리뷰 받고 싶은 포인트 (Review Focus)**

- 전체적인 흐름이 잘 잡히지 않습니다! 어떻게 잡으면 좋을지 조언부탁드립니다.

src/pages/communitypage.tsx

# 주강사 답변

페이지네이션을 구현할 때 당연하게 마주할 질문입니다.  
검색, 필터, 정렬기능을 위한 상태관리는 어떻게 해야하는가?  
새로고침으로 인한 상태값 소실은 어떻게 해결해야하는가?  
저 기능들을 구현할 때 UX를 해치지 않을 방법은 무엇인가?

## 1. 검색, 필터, 정렬기능을 위한 상태관리

페이지네이션 리스트 조회의 경우 여러 queryString이 함께 전달됩니다.  
서버에서는 이 queryString의 유효성검사를 1차적으로 진행하고, 유효하지 않은 경우 기본값으로 대체합니다. (또는 에러처리할 경우도 있습니다.)  
그렇게 가공된 queryString들을 쿼리문으로 조합하여 데이터를 조회합니다.  
그리고 조회된 데이터를 응답 형식에 맞게 가공하여 반환합니다. (유저에게는 꼭 필요한 최소한의 데이터만 전달됩니다.)

위 흐름을 살펴보면 프런트에서 하는 역할은 데이터 필터에 필요한 모든 데이터를 한 번에 던져주기만 하면 됩니다.

현재 리스트 데이터 필터를 위해 필요한 데이터는

1. 검색어
2. 검색 유형
3. 정렬 기준
4. 페이지
5. 카테고리

이정도 되어보이네요. (페이지당 보이는 개수를 조절하는 limit가 포함되는 경우도 많습니다.)

한 번에 쓰이는 데이터들이기 때문에 한 곳에서 관리를 해주면 실제 기능구현 측면에서도, 유지보수 측면에서도 도움이 될겁니다.

현재 코드를 보면

```tsx
function CommunityPage() {
  const { queryState } = useCommunityQuery();
```

필요한 모든 상태값들을 하나의 hook으로 관리하려는 시도를 볼 수 있습니다.  
좋은 구조입니다. 현업에서도 저렇게 구현하는 경우가 많습니다.  
물론 useCommunityQuery의 동작에 대한 개선은 조금 필요하지만 구조 자체는 괜찮은 상태입니다.

<details>
  <summary><strong>useCommunityQuery의 동작 개선 숙제</strong></summary>
  현재 코드는 구조 자체는 좋습니다. 어떻게 데이터 처리되는지 흐름은 거의 잘 맞습니다.

하지만 현업 기준에서 저 코드를 그대로 사용하기에는 타입 안정성의 문제가 있습니다.

### 1. 타입 안정성 확보

예를들면

```tsx
category_id: Number(searchParams.get('category_id')) || undefined,
```

이 코드는 안전한 타입 변환을 보장하지 못합니다. 그리고

```tsx
(searchParams.get('search_filter') as CommunitySearchFilter) ||
  undefined(searchParams.get('sort') as CommunitySearchSort) ||
  CommunitySearchSort.LATEST;
```

위 코드들은 enum 범위 밖 값이 와도 그대로 캐스팅됩니다. 런타임에서는 전혀 검증되지 않게 됩니다.  
위같은 코드들은 상태값의 일관성을 해치고 그로인해 UX를 해치게 될 수 있습니다.

### 2. 리셋 함수 개선

```tsx
const resetFilters = useCallback(() => {
  setSearchParams({});
}, [setSearchParams]);
```

모든 쿼리스트링을 제거하는 의도는 명확합니다.  
하지만 리셋하면서 유지해야할 기본값이 있는 경우가 많습니다.  
지금도 page값 같은 것들은 기본값을 유지하는 것이 안전합니다.

</details>

## 2. 새로고침으로 인한 상태값 소실

단순히 state로만 관리하는 것은 새로고침으로 인한 상태값 소실을 일으킬 수 있습니다.

웹 프런트에서 이 문제를 해결하는 가장 일반적인 방법은 URL(query string)을 사용하는 것입니다.

검색어, 필터, 정렬, 페이지 등의 상태를 URL에 반영하면,

- 새로고침을 해도 동일한 상태 유지 가능
- 뒤로 가기 / 앞으로 가기 히스토리 자연스럽게 동작
- URL 공유만으로 동일한 화면 재현 가능

이라는 장점이 있습니다.

그래서 단순 state관리가 아닌 지금같은 URL 기반 상태 관리가 더 적합한 선택입니다.

현재 useCommunityQuery는 이 방향을 잘 선택했고,  
타입 안정성과 기본값 처리만 보완되면 실무에서도 충분히 사용 가능한 구조입니다.

## 3. UX를 해치지 않을 방법은 무엇인가?

보통 이제 검색어 입력하고 입력된 값을 api로 보낼 때 문제가 발생할 수 있습니다.  
예를들어 검색어가 입력될 때마다 바로바로 api를 호출한다면

1. 네트워크 요청이 과도하게 발생하고
2. 리스트가 계속 깜빡이거나
3. 서버 응답 순서에 따라 의도치 않은 결과가 보일 수 있습니다.

실제로 3번은 네트워크, 서버의 상황에 따라 응답 순서가 바뀔 수 있습니다.  
k -> ki -> kim 순서로 입력해 요청을 보내더라도  
k -> kim -> ki 순서로 응답이 올 수 있습니다.  
이는 UX측면에서 심각한 문제가 됩니다. 서비스에 대한 유저의 신뢰도가 급하락하는 요인이 되죠.

하지만 이 부분은 이미 reactQuery 적용중이고 queryKey를 통해 방어하고 있는 부분으로 보입니다.

그렇기에 검색어 입력에서만 디바운스(Debounce)를 적용하면 충분할 것 같습니다.

디바운스는 입력값이 변할 때마다 즉시 API 호출하는 것이 아닌 딜레이(보통 300ms ~ 500ms 정도)를 두고 API 호출하는 기법입니다.

디바운스를 적용할 때 주의할 점은 모든 상태에 적용하는 것이 아니라, "의도적으로 입력이 잦은 상태"에만 적용하는 것입니다.

대표적으로:

- 검색어(search)
- 자유 입력 기반 필터

반면,

- 정렬 기준
- 카테고리
- 페이지 이동

같은 경우는 유저의 명확한 의도가 담긴 액션이기 때문에
즉시 API 호출이 오히려 더 자연스러운 UX가 됩니다.

즉 정리하면,

- 검색어 → 디바운스 적용
- 필터 / 정렬 / 페이지 → 즉시 반영

이렇게 역할을 나누는 것이 좋습니다.


## [추가 설명] URL 기반 상태 관리 추상화
검색, 필터, 정렬, 페이지네이션처럼 query string 기반 상태 관리는 특정 페이지에만 국한되지 않고 리스트 페이지가 있는 곳이라면 거의 항상 반복해서 사용됩니다.

이런 페이지마다 
> useSearchParams → 파싱 → 기본값 처리 → 업데이트

로직을 각각 구현하게 되면
- 코드 중복이 심해지고
- 타입 처리 방식이 페이지마다 달라지며
- 유지보수가 어려워지는 문제가 생깁니다.

그래서 실무에서는  
URL을 하나의 상태 저장소로 보고, 이를 다루는 **공통 추상 레이어를** 만드는 방식을 많이 사용합니다.

### 공통 훅(useUrlState)의 역할

아래와 같은 형태의 훅은 URL을 React state처럼 다룰 수 있게 만들어 줍니다.
```tsx
const { queryState, updateUrl, resetUrl } = useUrlState(defaultValues);
```

이 훅이 해결해주는 문제는 크게 세 가지입니다.

### 1. URL → 상태 파싱 로직의 중앙화
```tsx
const queryState = useMemo(() => { ... });
```
1. URL의 query string을 읽어서
2. 기본값(defaultValues)을 기준으로
3. number / boolean / string 타입 변환을 공통 규칙으로 처리

함으로써 각 페이지에서 직접 파싱 로직을 작성하지 않아도 되고, URL 상태 해석 기준이 프로젝트 전체에서 일관되게 유지됩니다.

### 2. URL 업데이트 로직의 표준화
```tsx
updateUrl({ page: 2, search: 'kim' });
```
- undefined / null / 빈 값은 자동으로 제거
- string 변환 로직을 한 곳에서 관리
- replace / push 여부도 옵션으로 제어

함으로써 페이지마다 setSearchParams를 직접 만지는 실수를 줄이고, 의도한 상태 변경만 안전하게 반영할 수 있습니다.

### 3. "URL 기반 상태"와 "페이지 로직"의 분리
이 훅의 가장 큰 장점은 URL을 다루는 책임과 페이지의 비즈니스 로직을 분리할 수 있다는 점입니다.

```tsx
// Community 전용 훅
const defaultQuery = {
  page: 1,
  search: '',
  sort: 'latest',
  category_id: undefined,
};

const { queryState, updateUrl } = useUrlState(defaultQuery);
```

- useUrlState → URL 관리 전담
- useCommunityQuery → 도메인(커뮤니티) 전용 규칙

이렇게 공통 로직 위에 페이지별 규칙을 얹는 구조가 됩니다.

# 결론

현재 구현 방향은 문제를 정확히 인식하고, 올바른 해결 방향을 선택한 상태로 보입니다.

- 검색, 필터, 정렬, 페이지네이션을 URL(query string)을 단일 상태 소스로 관리하려는 접근은 매우 적절합니다.
- React Query를 통해 요청 상태와 응답 순서 문제를 이미 상당 부분 해결하고 있으며,
- 검색어 입력에만 디바운스를 적용하는 방식 역시 UX와 성능 측면에서 합리적인 선택입니다.

하지만 실무 기준에서 더 안정적인 구조를 만들기 위해서는 다음 두 가지 보완이 핵심 포인트입니다.

1. query string → 상태 변환 과정에서의 타입 안정성 확보
   - 숫자 변환, enum 값 검증, 기본값 처리 로직을 명확히 분리

2. 리셋 및 기본값 정책을 명확히 정의
   - 모든 값을 지우는 것이 아닌, 유지되어야 할 기본 상태를 의도적으로 관리

이 두 가지만 정리된다면 현재 구조는 확장성과 유지보수성 모두를 고려한 좋은 페이지네이션 설계라고 볼 수 있습니다.

전체적인 흐름을 잡지 못했다고 느꼈을 수 있지만, 코드와 고민의 방향을 보면 이미 올바른 설계 궤도 위에 올라와 있는 상태입니다.

지금은 구조를 다시 뒤엎기보다는,  
기존 방향을 기준으로 "안정성"과 "명확성"을 보완해 나가는 단계라고 생각하시면 좋겠습니다.

SowonHub committed:
add alias

SowonHub committed:
add barrel pattern

SowonHub committed:
delete

SowonHub committed:
import barrel pattern

SowonHub committed:
edit date utils

SowonHub committed:
edit export default

SowonHub committed:
add ui

SowonHub committed:
edit components of community-page

SowonHub committed:
edit types-interface

SowonHub committed:
add mock list

SowonHub committed:
add handlers

SowonHub committed:
add cn

SowonHub committed:
add useAuthStore

SowonHub committed:
add community utils

SowonHub committed:
delete

SowonHub committed:
add husky

SowonHub committed:
edit paths

SowonHub committed:
add routes

SowonHub committed:
edit interface file

SowonHub committed:
edit communityApi

SowonHub committed:
edit routes

SowonHub committed:
edit categorymenu

SowonHub committed:
edit pagination

SowonHub committed:
edit postcard ui ( item -> card )

SowonHub committed:
edit ui ( select -> nativeselect )

SowonHub committed:
edit hook

SowonHub committed:
edit searchinput

SowonHub committed:
add ui

SowonHub committed:
edit handlers

SowonHub committed:
edit communitypage

SowonHub committed:
edit routes

SowonHub committed:
edit routes, (img->image), //

SowonHub committed:
delete

SowonHub committed:
add barrel pattern

SowonHub committed:
edit App

SowonHub committed:
edit routes

SowonHub committed:
delete

SowonHub committed:
edit api

SowonHub committed:
edit components

SowonHub committed:
edit ui

SowonHub committed:
edit simple

SowonHub committed:
edit handlers

SowonHub committed:
edit communitypage

SowonHub committed:
edit communitytype

SowonHub committed:
edit categori_id error

SowonHub committed:
refactor: reorganize API and types structure to match team conventions

- Move communityPostsApi.ts → api/post/services.ts
- Move communityPageTypes.ts → api/post/types.ts
- Change to named export (postService) for consistency with commentService
- Update all import paths to use new structure
- Fix CategoryId type imports (type CategoryId vs ECategoryId enum)
- Align with team's domain-based folder structure (api/comment, api/post)

Benefits:
- Better organization by domain
- API and types colocated
- Consistent naming conventions
- Easier to scale and maintain

SowonHub committed:
feat: 검색기능 구현(#27)

- vercel.json 추가
- URL 수정
- displayCategory 수정
- textare에 title, placeholder 추가
- getCategoryNameById 추가

SowonHub committed:
add all

SowonHub committed:
delete

SowonHub committed:
delete 2
